File: app/src/test/java/com/kickstarter/viewmodels/SurveyHolderViewModelTest.java
package com.kickstarter.viewmodels;

import com.kickstarter.KSRobolectricTestCase;
import com.kickstarter.libs.Environment;
import com.kickstarter.mock.factories.SurveyResponseFactory;
import com.kickstarter.models.Project;
import com.kickstarter.models.SurveyResponse;

import org.junit.Test;

import androidx.annotation.NonNull;
import rx.observers.TestSubscriber;

public class SurveyHolderViewModelTest extends KSRobolectricTestCase {
  private SurveyHolderViewModel.ViewModel vm;
  private final TestSubscriber<String> creatorAvatarImageUrl = new TestSubscriber<>();
  private final TestSubscriber<String> creatorNameTextViewText = new TestSubscriber<>();
  private final TestSubscriber<Project> projectForSurveyDescription = new TestSubscriber<>();
  private final TestSubscriber<SurveyResponse> startSurveyResponseActivity = new TestSubscriber<>();

  private void setUpEnvironment(final @NonNull Environment environment) {
    this.vm = new SurveyHolderViewModel.ViewModel(environment);

    this.vm.outputs.creatorAvatarImageUrl().subscribe(this.creatorAvatarImageUrl);
    this.vm.outputs.creatorNameTextViewText().subscribe(this.creatorNameTextViewText);
    this.vm.outputs.projectForSurveyDescription().subscribe(this.projectForSurveyDescription);
    this.vm.outputs.startSurveyResponseActivity().subscribe(this.startSurveyResponseActivity);
  }

  @Test
  public void testCreatorAvatarImageUrl() {
    final SurveyResponse surveyResponse = SurveyResponseFactory.surveyResponse();
    setUpEnvironment(environment());
    this.vm.inputs.configureWith(surveyResponse);
    this.creatorAvatarImageUrl.assertValues(surveyResponse.project().creator().avatar().small());
  }

  @Test
  public void testCreatorNameEmits() {
    final SurveyResponse surveyResponse = SurveyResponseFactory.surveyResponse();
    setUpEnvironment(environment());
    this.vm.inputs.configureWith(surveyResponse);
    this.creatorNameTextViewText.assertValues(surveyResponse.project().creator().name());
  }

  @Test
  public void testSurveyDescription() {
    final SurveyResponse surveyResponse = SurveyResponseFactory.surveyResponse();
    setUpEnvironment(environment());
    this.vm.inputs.configureWith(surveyResponse);
    this.projectForSurveyDescription.assertValues(surveyResponse.project());
  }

  @Test
  public void testStartSurveyResponseActivity() {
    final SurveyResponse surveyResponse = SurveyResponseFactory.surveyResponse();
    setUpEnvironment(environment());

    this.vm.inputs.configureWith(surveyResponse);
    this.vm.inputs.surveyClicked();

    this.startSurveyResponseActivity.assertValue(surveyResponse);
  }
}

File: app/src/test/java/com/kickstarter/viewmodels/SurveyHolderViewModelTest.kt
package com.kickstarter.viewmodels

import com.kickstarter.KSRobolectricTestCase
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.mock.factories.SurveyResponseFactory.surveyResponse
import com.kickstarter.models.Project
import com.kickstarter.models.SurveyResponse
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subscribers.TestSubscriber
import org.junit.After
import org.junit.Test

class SurveyHolderViewModelTest : KSRobolectricTestCase() {

    private lateinit var vm: SurveyHolderViewModel.ViewModel

    private val creatorAvatarImageUrl = TestSubscriber<String>()
    private val creatorNameTextViewText = TestSubscriber<String>()
    private val projectForSurveyDescription = TestSubscriber<Project>()
    private val startSurveyResponseActivity = TestSubscriber<SurveyResponse>()
    private val disposable = CompositeDisposable()

    private fun setUpEnvironment() {
        vm = SurveyHolderViewModel.ViewModel()
        vm.outputs.creatorAvatarImageUrl().subscribe { creatorAvatarImageUrl.onNext(it) }
            .addToDisposable(disposable)
        vm.outputs.creatorNameTextViewText().subscribe { creatorNameTextViewText.onNext(it) }
            .addToDisposable(disposable)
        vm.outputs.projectForSurveyDescription()
            .subscribe { projectForSurveyDescription.onNext(it) }.addToDisposable(disposable)
        vm.outputs.startSurveyResponseActivity()
            .subscribe { startSurveyResponseActivity.onNext(it) }.addToDisposable(disposable)
    }

    @Test
    fun testCreatorAvatarImageUrl() {
        val surveyResponse = surveyResponse()
        setUpEnvironment()

        vm.inputs.configureWith(surveyResponse)

        creatorAvatarImageUrl.assertValues(surveyResponse.project()?.creator()?.avatar()?.small())
    }

    @Test
    fun testCreatorNameEmits() {
        val surveyResponse = surveyResponse()
        setUpEnvironment()

        vm.inputs.configureWith(surveyResponse)

        creatorNameTextViewText.assertValues(surveyResponse.project()?.creator()?.name())
    }

    @Test
    fun testSurveyDescription() {
        val surveyResponse = surveyResponse()
        setUpEnvironment()

        vm.inputs.configureWith(surveyResponse)

        projectForSurveyDescription.assertValues(surveyResponse.project())
    }

    @Test
    fun testStartSurveyResponseActivity() {
        val surveyResponse = surveyResponse()

        setUpEnvironment()

        vm.inputs.configureWith(surveyResponse)
        vm.inputs.surveyClicked()

        startSurveyResponseActivity.assertValue(surveyResponse)
    }

    @After
    fun clear() {
        disposable.clear()
    }
}
